### ISSUE TITLE
Error highlighting does not work

### ISSUE THREAD

**Issue by whysoserious** [clojure-emacs/cider#1175]
#### Steps to reproduce

* Create any lein project eg. `lein new quil my-sketch`
* Open any `.clj` file eg. `src/my_sketch/core.clj`
* `M-x cider-jack-in`
* Make any syntax error which should be highlighted
* `M-x cider-eval-defun-at-point` - __yields no highlighting__
* `M-x cider-eval-region` - __yields no highlighting__
* `M-x cider-eval-buffer` - works as expected - I can see red underscores

##### My emacs config: 

https://github.com/whysoserious/my-emacs-config/blob/master/init.el#L326-L374

##### `M-x version`

```
GNU Emacs 24.4.1 (x86_64-apple-darwin13.4.0, NS apple-appkit-1265.21) of 2014-10-21 on builder10-9.porkrind.org
```

##### repl

```
; CIDER 0.9.1 (package: 20150625.946) (Java 1.8.0_25, Clojure 1.6.0, nREPL 0.2.10)
WARNING: The following nREPL ops are not supported:
extract-definition rename-file-or-dir stubs-for-interface warm-ast-cache
Please, install (or update) refactor-nrepl and restart REPL.
You can mute this warning by changing cljr-suppress-middleware-warnings.
WARNING: clj-refactor and refactor-nrepl are out of sync.
Their versions are 1.1.0-SNAPSHOT and 1.0.5, respectively.
You can mute this warning by changing cljr-suppress-middleware-warnings.
user>
```

**Comment by Malabarba** [comment]
There are two things going on here. Let's tackle them 1 at a time:

### The `eval` op doesn't check for syntax errors until execution

If you write something like the following, and evaluate it with `cider-eval-defun`, the nrepl will simply not throw any errors.

    (defn should-error []
      (reduce))

There's not much cider can do, the `eval` op simply doesn't check for errors (I think it doesn't compile the function?). It's not just about highlighting, you don't even get a stacktrace buffer. If you issue `load-buffer` instead, then you *will* get a stacktrace and highlighting, because the `load-file` op *does* compile.

In order to trigger an error, you have to actually use the function, for instance by evaluating `(should-error)`. Then you will get an error and a stacktrace, but you still don't get highlighting. That's the second problem.

### When execution runs into a syntax error, the eval op produces an uninformative error.

Right now, highlighting is done entirely by calling `cider-extract-error-info` on the error message sent by the REPL. That is, it has nothing to do with the actual stacktrace buffer (they just happen to always be called together).
To add to the problem, the error message the this function gets is very short and uninformative in the case of interactive eval:

     "ArityException Wrong number of args (1) passed to: core/reduce  clojure.lang.AFn.throwArity (AFn.java:429)\n"

Whereas the message in the case of “load-file” is much more informative:

    "CompilerException clojure.lang.ArityException: Wrong number of args (1) passed to: core/reduce, compiling:(/home/am12548/Git-Projects/clojure-playground/src/clojure_playground/core.clj:17:30) \n"

In order to fix this we would need to reword `cider-extract-error-info` a little, and then call it as part of the same function that generates the stacktrace, so that this function could give it a little more information to work with.

**Comment by bbatsov** [comment]
> There's not much cider can do, the eval op simply doesn't check for errors (I think it doesn't compile the function?). It's not just about highlighting, you don't even get a stacktrace buffer. If you issue load-buffer instead, then you will get a stacktrace and highlighting, because the load-file op does compile.
> 
> In order to trigger an error, you have to actually use the function, for instance by evaluating (should-error). Then you will get an error and a stacktrace, but you still don't get highlighting. That's the second problem.

I'm pretty confused by this command. `eval` actually compiles the code and reports compilation errors just as `load-file` does. The problem is that eval doesn't attach proper locations to the stacktraces frames, because as far as it's concerned they happened in the temp (tooling) file used to eval (compile) the code in question.

> "ArityException Wrong number of args (1) passed to: core/reduce  clojure.lang.AFn.throwArity (AFn.java:429)\n"

Hmm, that doesn't sound quite right. I recall the messages were pretty much the same.

**Comment by Malabarba** [comment]
@bbatsov 
> I'm pretty confused by this command. eval actually compiles the code and reports compilation errors just as load-file does. The problem is that eval doesn't attach proper locations to the stacktraces frames, because as far as it's concerned they happened in the temp (tooling) file used to eval (compile) the code in question.

If I evaluate the defn above with any of the eval commands, I get *no* errors and no stacktrace frames. I set `(setq nrepl-log-messages t)` and I verified that the only messages I get back form the nrepl are the "done" status and the return value (which is `#'should-error`, as if everything were fine). Is that not what you get?

My understanding (from looking at the source of `interruptible_eval.clj`) is that it simply calls the `eval` function on the code.

**Comment by bbatsov** [comment]
But `eval` in Clojure is actually `compile`. There's no evaluation, only compilation.

**Comment by Malabarba** [comment]
Strange. But I still don't get errors when evaluating that. I also don't get errors if I paste that in a regular `lein repl` (not sure if it's related).

**Comment by bbatsov** [comment]
The definition or its invocation? For me evaluating invalid code yields the expected results:

```
(shit)

  err  "CompilerException java.lang.RuntimeException: Unable to resolve symbol: shit in this context, compiling:(/private/var/folders/f0/cm24s0qd0s5cch5jvj6ctnb00000gn/T/form-init8023551865063802074.clj:1:1) \n"
  id  "8"
  session  "3253b0d0-386c-4169-8936-c1181f8e9924"
```

**Comment by bbatsov** [comment]
Hmm, you're right - with your example there are no precise code locations. Likely a Clojure bug...

**Comment by Malabarba** [comment]
No. I'm taking about the definition. Evaluating a definition that contains invalid code doesn't yield errors. But doing load-file on the file that contains that definition *does* yield an error.

Evaluating the invocation does yield an error (that's the second part of my comment).

**Comment by Malabarba** [comment]
I'm taking about both separately.

1. Evaluating a definition that contains invalid code doesn't yield errors. But doing load-file on the file that contains that definition *does* yield an error.
2. Evaluating the invocation does yield an error (that's the second part of my comment).

**Comment by bbatsov** [comment]
> Evaluating a definition that contains invalid code doesn't yield errors. But doing load-file on the file that contains that definition does yield an error.

Hmm, that definitely sounds like some bug. //cc @cemerick 

> Evaluating the invocation does yield an error (that's the second part of my comment).

Although this error lacks column info for some reason.

**Comment by Malabarba** [comment]
> Although this error lacks column info for some reason.

Frankly, that's the least of our problems. You'll note that the file info is simply wrong. It points to this `AFn.java` file, instead of pointing to a tooling file (which is where the function was defined).

    ArityException Wrong number of args (0) passed to: core$reduce  clojure.lang.AFn.throwArity (AFn.java:437)

**Comment by bbatsov** [comment]
This probably has to be reported upstream. If this is not a bug I don't know what it is. :-)

**Comment by Malabarba** [comment]
Where do you report clojure bugs? :-)

**Comment by bbatsov** [comment]
http://dev.clojure.org/jira/browse/CLJ

**Comment by Malabarba** [comment]
Actually, let me rectify. I was doing something stupid before, and there's actually no difference for me between `eval` or `load-file`. Neither of them throw errors when defining a "bad" function. Both only throw errors when the function is actually called.

**Comment by bbatsov** [comment]
Yeah, I had started writing you a reply regarding this at some point during the day, but for some reason I haven't send it.

**Comment by Malabarba** [comment]
@bbatsov I think this discussion buried the original issue a little. I've edited my original comment with what I think is right now. Would you like to delete some of extra comments?

**Comment by bbatsov** [comment]
Well, the missing locations should be fixed in Clojure for people using nREPL 0.2.11+ and we can't really fix the related Clojure bug, so I'll close this.
